### PR TITLE
fix: remove trailing slash for invalid URL

### LINKED ISSUE #395: [BUG]: Undefined slash-parameters likely expanded incorrectly
### What happened?

```
> const { endpoint } = require("@octokit/endpoint");
> endpoint("http://api.github.com/repos/octocat/Hello-World/branches{/branch}", {})
{
  method: 'GET',
  url: 'http://api.github.com/repos/octocat/Hello-World/branches/',
  headers: {
    accept: 'application/vnd.github.v3+json',
    'user-agent': 'octokit-endpoint.js/7.0.4 Node.js/19.4.0 (linux; x64)'
  }
}
```

Note the slash at the end of `url`. GitHub returns 404 for `.../branches/`, but lists branches for `.../branches`.

Expanding the same URL with `url-template` and empty context returns `.../branches`. This makes me belive that `endpoint.js` performs template expansion incorrectly.

### Versions

Octokit.js v7.0.4, Node v19.4.0

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by maemaemae3**
Resolves #395

----

### Before the change?

* remove trailing slashes from URL except expanded endpoint is `/`

### After the change?

* if you specify url like `endpoint("https://github.acme-inc/api/v3/")`, it will cause to trailing slash removed.

### Pull request checklist
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)

### Does this introduce a breaking change?

* if you specify url like `endpoint("https://github.acme-inc/api/v3/")`, it will cause to trailing slash removed.

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/master/community/breaking_changes.md) to help!

- [x] Yes
- [ ] No

----

### CODE REVIEW

**Comment by maemaemae3** [comment]
considering for effects about breaking changes.
It seems there's no issue with omitting the trailing slash, but I'm considering whether there could be any unexpected effects.

**Review by gr2m — COMMENTED** [review]
I like the change, I've seen lots of people reporting issues that turned out to be a trailing slash. But we have to be careful to not accidentally break use cases. For example, can you please test with https://docs.github.com/en/free-pro-team@latest/rest/git/refs?apiVersion=2022-11-28#list-matching-references and https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#get-repository-content (loading a folder)

I think in either case the `/` should be URL encoded, but I'm not sure

**Comment by maemaemae3** [comment]
I added tests for trailing slash in expressions, and checked its encoded properly.
However trailing slash in https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#get-repository-content (loading a folder) returns `Bad request`, which means you must omit trailing slash.
(I mean API like `http://api.github.com/repos/my-org/my-repo/contents/mydir%2Fnested-dir%2F` returns Bad request.)
Should we also consider cases like this?

File: test/endpoint.test.ts [hunk 1]
```diff
@@ -500,6 +500,72 @@ describe("endpoint()", () => {
...
+    const options2 = endpoint("GET /repos/{owner}/{repo}/contents/{path}", {
+      owner: "owner",
+      repo: "repo",
+      path: "heads/",
+      ref: "feature/branch",
+    });
+
+    expect(options2).toEqual({
+      method: "GET",
+      url: "https://api.github.com/repos/owner/repo/contents/heads%2F?ref=feature%2Fbranch",
```
Comment (gr2m) [inline test/endpoint.test.ts:561]:
```suggestion
      path: "my-folder/",
      ref: "feature/branch",
    });

    expect(options2).toEqual({
      method: "GET",
      url: "https://api.github.com/repos/owner/repo/contents/my-folder%2F?ref=feature%2Fbranch",
```

**Comment by gr2m** [comment]
> However trailing slash in https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#get-repository-content (loading a folder) returns `Bad request`, which means you must omit trailing slash

That is something we should fix on the API side, so folks not using Octokit will benefit as well. I'll raise that internally.

**Review by gr2m — APPROVED** [review]
I am a bit worried about unforeseen consequences. I'm sure there are users that use octokit to send requests to non-github endpoints and this change might break their code. But I would not take that into account as well, it will bloat `@octokit/endpoint` even more, which is our lowest-level @octokit library for JavaScript.

I my vote is to merge, release, and see if problems pop up. It's easy enough to reverse

**Comment by kfcampbell** [comment]
:+1: I'll merge this now and standby to rollback if necessary.

**Pull request merged**